### PR TITLE
BUG: fix bug causing classes to not be vectorizable, closes #603

### LINKED ISSUE #603: Discretization of a functional fails
Noticed when working on pull #602 and trying to discretize a functional, we get a weird error:

```
>>> r2 = odl.rn(2)
>>> l2norm = odl.solvers.L2Norm(r2)
>>> space = odl.uniform_discr([-1, -1], [3, 3], [5, 5])
>>> space.element(l2norm, vectorized=False)

  File "d:\programming\odl\odl\discr\lp_discr.py", line 317, in element
    inp_elem = self.uspace.element(inp, vectorized=vectorized)
  File "d:\programming\odl\odl\space\fspace.py", line 667, in element
    fcall = vectorize(otypes=[dtype])(fcall)
  File "d:\programming\odl\odl\util\vectorization.py", line 179, in __call__
    return self._wrapper(func, *self.wrapper_args, **self.wrapper_kwargs)
  File "d:\programming\odl\odl\util\vectorization.py", line 232, in _wrapper
    **vect_kwargs))
  File "D:\Anaconda\lib\functools.py", line 33, in update_wrapper
    setattr(wrapper, attr, getattr(wrapped, attr))
AttributeError: 'L2Norm' object has no attribute '__name__'
```

Wrapping it in a function gives another problem, seemingly unrelated, everything becomes NaN

```
>>> def f(x): l2norm(x)
>>> space.element(f, vectorized=False)
uniform_discr([-1.0, -1.0], [3.0, 3.0], [5, 5]).element([[nan, nan, nan, nan, nan],
 [nan, nan, nan, nan, nan],
 [nan, nan, nan, nan, nan],
 [nan, nan, nan, nan, nan],
 [nan, nan, nan, nan, nan]])
```

@kohr-h is kindoff in charge of the vectorization, so perhaps you know why this happens?

Edit: Change to use a more standard functional. Error stays.

Edit2: Second case lacks a return statement, my fault, but perhaps.we should throw here?

### PR DESCRIPTION
**Pull request by adler-j**
Fixed #603 and added a test for it

### CODE REVIEW

File: odl/util/vectorization.py [hunk 1]
```diff
@@ -228,6 +228,10 @@ class vectorize(OptionalArgDecorator):
     @staticmethod
     def _wrapper(func, *vect_args, **vect_kwargs):
         """Return the vectorized wrapper function."""
+        if not hasattr(func, '__name__'):
+            # Set name if not available. Happens if func is actually a callable
+            func.__name__ = '{}.__call__'.format(func.__class__.__name__)
```
Comment (kohr-h) [inline odl/util/vectorization.py]: So same question here: why append `.__call__` to the name? Technically, neither `Class.__call__` nor `Class` will be usable directly in the wrapper, hence appending `__call__` does not really make the printed construction work literally.

File: test/util/vectorization_test.py [hunk 2]
```diff
@@ -315,5 +315,40 @@ def simple_func(x):
     assert simple_func(val_2) == 1
 
 
+def test_vectorize_callable_class():
+
+    # Test vectorization in 1d without data type --> lazy vectorization
+    arr = (np.arange(5) - 2)[None, :]
+    mg = sparse_meshgrid(np.arange(5) - 3)
```
Comment (kohr-h) [inline test/util/vectorization_test.py]: Use `[-3, -2, -1, 0, 1]`, easier to read.

File: test/util/vectorization_test.py [hunk 3]
```diff
@@ -315,5 +315,40 @@ def simple_func(x):
     assert simple_func(val_2) == 1
 
 
+def test_vectorize_callable_class():
+
+    # Test vectorization in 1d without data type --> lazy vectorization
+    arr = (np.arange(5) - 2)[None, :]
```
Comment (kohr-h) [inline test/util/vectorization_test.py]: Hard to see the slight difference between the two ranges. Better use `[-2, -1, 0, 1, 2]` explicitly.

**Comment by adler-j** [comment]
Fixed the minor stuff, will merge after CI.

**Pull request merged**